### ISSUE TITLE
Urock: high velocity streaks and windward vortices

### ISSUE THREAD

**Issue by bweeding** [UMEP-dev/UMEP-processing#53]
When running Urock, high velocity streaks and vortices on the windward side of a large structure appear. The sample files below were run with a 10m/s wind at 10m height, 315 degrees, and an output height of 1.5m. Horizontal and vertical resolutions were both 2m. The peak wind speed in the image below is ~57m/s.

![image](https://github.com/UMEP-dev/UMEP-processing/assets/57697604/3bdede04-c613-4e82-b5e4-8122b0a1716d)

[urock_outputs.zip](https://github.com/UMEP-dev/UMEP-processing/files/13551977/urock_outputs.zip)
[urock_source_files_issue.zip](https://github.com/UMEP-dev/UMEP-processing/files/13552065/urock_source_files_issue.zip)

**Comment by j3r3m1** [comment]
Hi @bweeding,

> - the one you are pointing might be due to the wind solver (second step of the Röckle method). We have a kind of similar problem for edges of courtyard, the wind flow seems not balanced, we loose a part of the flow within the buildings... I have not spotted the problem yet.

The problem has been solved for an other issue but it did not solve your. With more investigation the problem came from the street canyon scheme which is activated whenever the direction of the street canyon. In your case, the street canyon was activated with wind direction going toward the coming upwind direction. I have set an activation of the street canyon scheme only if the street canyon is less than 60° from the perpendicular to the wind direction (https://github.com/UMEP-dev/UMEP-processing/commit/1420ae22499bfab955f485d1ade836439775c477).

> - the second is the wind patterns in the upstream of the cylinder building. This comes from the fact that we define a single upstream zone for each polygon edge. It is fine for kind of "rectangular" shape buildings but rounded shapes create this problem. I think I see how to solve that, I will try to merge the upstream zones whenever it is needed as we did for the cavity zones.

Solved here: https://github.com/UMEP-dev/UMEP-processing/commit/1420ae22499bfab955f485d1ade836439775c477

Let us know if works fine.

**Comment by bweeding** [comment]
Hi @j3r3m1 , I've updated and rerun with the same settings. Things have improved, with the max velocity now down at 22m/s, but still seeing some streaking. The leading edge vortex pattern has gone :)

Any thoughts about dealing with the streaks? Perhaps I could apply a smoothing or change limit filter?

![image](https://github.com/UMEP-dev/UMEP-processing/assets/57697604/b270046e-056d-4310-abd2-5a96b164e823)

**Comment by j3r3m1** [comment]
OK, good catch again.

This one is due to the current limitation of the URock way of dealing with the wake zones.
The string of high wind speed is located in the cavity zone of the closest upwind building while it is surrounded by cavity zone of the big cylinder building and by backward cavity zones and street canyons.
![image](https://github.com/UMEP-dev/UMEP-processing/assets/13120405/c489b985-a104-4f34-9d8e-db0a2f432a2a)

This leads to the following initialization:
![image](https://github.com/UMEP-dev/UMEP-processing/assets/13120405/aa110b7a-9deb-4efd-bbbe-b0f99e63aebb)

As you can see, only the wind vectors coming from the cavity zones are visible. The south-west ones belong to the closest building while the north-east ones belong to the cylinder. The south-west wind speed should normally be lower but this is one limitation of the way URock is dealing with superimpositions.

Below are two buildings, wind comes from left. The cavity zone of the buildings are in blue and the wake zone in orange. 

![image](https://github.com/UMEP-dev/UMEP-processing/assets/13120405/c270d502-c4ad-41d5-bf9b-f2cdb1bbf1ca)

This case fits with the current south-west arrows above: the upstream building creates itw own cavity zone. The problem is that its values should be decreased by the wake zone of the cylinder building. It is the case in URock but only for cells being above the cylinder cavity zone. The cells located below the cavity zone line of the cylinder should also be affected in this specific case since the top of the smallest building is in the wake zone of the cylinder building. However, it is not the case. This point should be improved in the future but unfortunately I do not see a simple way to do that right now.

**Comment by bweeding** [comment]
Ah ok, an interesting problem! I'm planning to use Urock in a coming paper, so I will have to go with some kind of filtering or smoothing to address the issue while acknowledging it. Do you have any suggestions?

Also, if I run the simulation using the urban wind profile (rather than power), then the magnitude is far lower again. Maximum of 8.1m/s:
![image](https://github.com/UMEP-dev/UMEP-processing/assets/57697604/09c48fc1-ab91-45c3-83b5-8e9a386a1377)

**Comment by j3r3m1** [comment]
Yes the reason for the much lower wind speed using urban wind profile is that the wind speed increases much slower with height in this case. The reference wind speed for many Röckle zones being the one at the top of the building, the wind profile used may have a large effect on the absolute wind speed at the bottom. However, the QUIC-URB developpers recommend to use the power law instead of the urban one. Ideally, if you have a wind profile for your zone this is even better.

A simple way to filter the anormally high wind speed would be to use a threshold of wind speed based on a rule that you would have observed for an important number of cases (neighbordhoods).

I would suggest that for each tile you have you identify:
- the highest wind speed of your anormal high wind speed zones 
- the highest wind speed of your normal high wind speed zones
- the mean wind speed of your zone
- the std wind speed of your zone

--> according to that try to find a parameter p that would exclude most of your anormal high wind speed values while keeping the normal ones. It might be one of:
- threshold = p * mean wind speed
- threshold = mean wind speed + p * std wind speed

**Comment by bweeding** [comment]
Ok cool, I'll investigate!

**Comment by bweeding** [comment]
Ah yep close away!
